tests/cpptests/test_cpp_iterator_intersection.cpp

```diff
@@ -0,0 +1,353 @@
+    IntersectionIterator *ii = (IntersectionIterator *)ii_base;
+    IteratorStatus rc;
+    // Test skipping to any id between 1 and the last id
+    t_docId i = 1;
+    for (t_docId id : resultSet) {
+        while (i < id) {
+            ii_base->Rewind(ii_base);
+            rc = ii_base->SkipTo(ii_base, i);
+            ASSERT_EQ(rc, ITERATOR_NOTFOUND);
+            ASSERT_EQ(ii->base.lastDocId, id);
+            ASSERT_EQ(ii->base.current->docId, id);
+            i++;
+        }
+        ii_base->Rewind(ii_base);
+        rc = ii_base->SkipTo(ii_base, id);
+        ASSERT_EQ(rc, ITERATOR_OK);
+        ASSERT_EQ(ii->base.lastDocId, id);
+        ASSERT_EQ(ii->base.current->docId, id);
+        i++;
+    }
+    // Test reading after skipping to the last id
+    ASSERT_EQ(ii_base->Read(ii_base), ITERATOR_EOF);
+    ASSERT_EQ(ii_base->SkipTo(ii_base, ii_base->lastDocId + 1), ITERATOR_EOF);
+    ASSERT_TRUE(ii->base.atEOF);
+
+    ii_base->Rewind(ii_base);
+    ASSERT_EQ(ii->base.lastDocId, 0);
+    ASSERT_FALSE(ii->base.atEOF);
+    // Test skipping to all ids that exist
+    for (t_docId id : resultSet) {
+        rc = ii_base->SkipTo(ii_base, id);
+        ASSERT_EQ(rc, ITERATOR_OK);
+        ASSERT_EQ(ii->base.lastDocId, id);
+        ASSERT_EQ(ii->base.current->docId, id);
+    }
+
+    // Test skipping to an id that exceeds the last id
+    ii_base->Rewind(ii_base);
+    ASSERT_EQ(ii->base.lastDocId, 0);
+    ASSERT_FALSE(ii->base.atEOF);
+    rc = ii_base->SkipTo(ii_base, resultSet.back() + 1);
+    ASSERT_EQ(rc, ITERATOR_EOF);
+    ASSERT_EQ(ii->base.lastDocId, 0); // we just rewound
+    ASSERT_TRUE(ii->base.atEOF);
+}
+
+TEST_P(IntersectionIteratorCommonTest, Rewind) {
+    IntersectionIterator *ii = (IntersectionIterator *)ii_base;
+    IteratorStatus rc;
+    for (int i = 0; i < 5; i++) {
+        for (int j = 0; j <= i; j++) {
+            ASSERT_EQ(ii_base->Read(ii_base), ITERATOR_OK);
+            ASSERT_EQ(ii->base.current->docId, resultSet[j]);
+            ASSERT_EQ(ii->base.lastDocId, resultSet[j]);
+        }
+        ii_base->Rewind(ii_base);
+        ASSERT_EQ(ii->base.lastDocId, 0);
+        ASSERT_FALSE(ii->base.atEOF);
+    }
+}
+
+// Parameters for the tests above. We run all the combinations of:
+// 1. number of child iterators in {2, 5, 25}
+// 2. expected result set, one of the 3 given lists below
+INSTANTIATE_TEST_SUITE_P(IntersectionIterator, IntersectionIteratorCommonTest, ::testing::Combine(
+    ::testing::Values(2, 5, 25),
+    ::testing::Values(
+        std::vector<t_docId>{1, 2, 3, 40, 50},
+        std::vector<t_docId>{5, 6, 7, 24, 25, 46, 47, 48, 49, 50, 51, 234, 2345, 3456, 4567, 5678, 6789, 7890, 8901, 9012, 12345, 23456, 34567, 45678, 56789},
+        std::vector<t_docId>{9, 25, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200, 210, 220, 230, 240, 250}
+    )
+));
+
+class IntersectionIteratorTest : public ::testing::Test {
```

Comment:
Isn't there a way to use mocks so that we do not need to create an InvertedIndex.